src/main/java/com/google/idea/perf/TracerView.kt

```diff
@@ -16,20 +16,86 @@
 
 package com.google.idea.perf
 
+import com.intellij.openapi.application.ApplicationManager.getApplication
 import com.intellij.openapi.ui.MessageType
 import com.intellij.openapi.ui.popup.util.PopupUtil
 import com.intellij.ui.components.JBLabel
 import com.intellij.ui.components.JBPanel
 import com.intellij.util.textCompletion.TextFieldWithCompletion
+import java.awt.event.KeyEvent
+import java.awt.event.KeyListener
 import javax.swing.JProgressBar
 
 /** The content for filling the tracer dialog window. */
 abstract class TracerView: JBPanel<TracerView>() {
+    companion object {
+        const val COMMAND_HISTORY_LIMIT = 20
+    }
+
+    abstract val controller: TracerController
     abstract val commandLine: TextFieldWithCompletion
     abstract val progressBar: JProgressBar
     abstract val refreshTimeLabel: JBLabel
 
+    private val commandHistory = ArrayList<String>(COMMAND_HISTORY_LIMIT)
+    private var commandHistoryIndex = -1
+
     fun showCommandBalloon(message: String, type: MessageType) {
         PopupUtil.showBalloonForComponent(commandLine, message, type, true, null)
     }
+
+    fun initEvents(controller: TracerController) {
+        val textComponent = commandLine.editor!!.contentComponent
+        textComponent.addKeyListener(object: KeyListener {
+            override fun keyTyped(e: KeyEvent) {}
+
+            override fun keyPressed(e: KeyEvent) {
+                if (e.keyCode == KeyEvent.VK_ENTER) {
+                    controller.handleRawCommandFromEdt(commandLine.text)
+
+                    if (commandHistoryIndex != -1) {
+                        if (commandHistory.size >= COMMAND_HISTORY_LIMIT) {
+                            commandHistory.removeAt(0)
+                        }
+                        while (commandHistory.lastIndex > commandHistoryIndex) {
+                            commandHistory.removeAt(commandHistory.lastIndex)
+                        }
+                    }
+
+                    if (commandHistory.isEmpty() || commandHistory.last() != commandLine.text) {
+                        commandHistory.add(commandLine.text)
+                        commandHistoryIndex = commandHistory.lastIndex
+                    }
+
+                    getApplication().invokeLater {
+                        commandLine.text = ""
```

Comment:
I think I actually missed this in a previous commit, but: why use `invokeLater` here? I expect this code is already called on the EDT.

Ditto for the two `invokeLater` calls below.

Reply:
You're right, removing the `invokeLater` calls still works. This was a leftover thought from the multiline hack when the `commandLine` component had to mutate itself in response to a document event listener.